### DISCUSSION TITLE
Different access tokens generated from different clients can be verified in same client.

### DISCUSSION THREAD

**Discussion by foogao** [keycloak/keycloak#23337] [Q&A]
Hello guys,

We have a use case as follows:

We want to create 2 keycloak clients in one Realm for 2 different login scenarios, the grant type is password. We meet the following issue:

The access token generated from client 1 can be verified by client 2. 

Do we have a way to isolate the access token with client? 

Our keycloak version is 22.0.1.

Thanks

**Comment by justin-tay** [comment]
The Resource Server is supposed to validate that the `access_token` has the appropriate claims as described at https://datatracker.ietf.org/doc/html/rfc9068#name-validating-jwt-access-token.

Typically for your use case the `aud` claim is used to determine if the Resource Server is the intended target of the `access token`.

If you really wish to differentiate the client that acquired the token, the `azp` or `client_id` claim can be used to identify that client.

Note that the `password` grant is typically not recommended see https://datatracker.ietf.org/doc/html/rfc6819#section-4.4.3.

**Reply by foogao** [reply]
Hi man, thanks a lot for your kindly response.

Question about the acronym `azp', what is it? 😄 

And the bearer JWT `access token` is generated/verified by the token endpoint, how can we control the claims of the token in the generation process?

Thanks.

**Reply by justin-tay** [reply]
The `azp` claim stands for Authorized Party and is part of the OpenID Specification. Rightfully you should see `azp` on `id_token` and `client_id` on `access_token` but currently Keycloak uses `azp` on `access_token` as well. See https://openid.net/specs/openid-connect-core-1_0.html#IDToken.

The tokens are generated by the token endpoint but validated by the Resource Servers that are receiving the tokens.

You customize the claims by using mappers as documented at https://www.keycloak.org/docs/latest/server_admin/#_protocol-mappers.
